### PR TITLE
ESQL: Fix a bug in VALUES agg

### PR DESCRIPTION
**Pull request by nik9000**
This fixes a bug in the VALUES agg when run on IP fields and grouped. Specifically, we could get array-out-of-bounds errors if the array in which we store the ips is oversized but not sized on the bounds of an IP. *AND* some group IDs received only `null` values - specifically those who were assigned the highest ordinals.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/AbstractArrayState.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
  *     Most of this class subclasses are autogenerated.
  * </p>
  */
-public class AbstractArrayState implements Releasable {
+public abstract class AbstractArrayState implements Releasable, GroupingAggregatorState {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/AbstractArrayState.java:22]: If I make this `abstract` and `implements` `GroupingAggregatorState` it's a little easier to poke it in tests.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java [hunk 2]
```diff
@@ -29,21 +38,37 @@ public static List<Object[]> params() {
...
-            params.add(new Object[] { DataType.FLOAT, 1000, inOrder });
-            params.add(new Object[] { DataType.DOUBLE, 1000, inOrder });
-            params.add(new Object[] { DataType.IP, 1000, inOrder });
+            for (IntSupplier count : new IntSupplier[] { new Fixed(100), new Fixed(1000), new Random(100, 5000) }) {
+                params.add(new Object[] { DataType.INTEGER, count, inOrder });
+                params.add(new Object[] { DataType.LONG, count, inOrder });
+                params.add(new Object[] { DataType.FLOAT, count, inOrder });
+                params.add(new Object[] { DataType.DOUBLE, count, inOrder });
+                params.add(new Object[] { DataType.IP, count, inOrder });
+            }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java:47]: More lengths makes it more likely we'll expose this bug. Including the randomized one.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java [hunk 3]
```diff
@@ -29,21 +38,37 @@ public static List<Object[]> params() {
...
+            return i;
+        }
+    }
+
+    private record Random(int min, int max) implements IntSupplier {
+        @Override
+        public int getAsInt() {
+            return randomIntBetween(min, max);
+        }
+    }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java:64]: These are `record`s so they have a nice `toString`.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java [hunk 4]
```diff
@@ -29,21 +38,37 @@ public static List<Object[]> params() {
...
+
     private final DataType type;
     private final ElementType elementType;
     private final int valueCount;
     private final boolean inOrder;
 
-    public ArrayStateTests(DataType type, int valueCount, boolean inOrder) {
+    public ArrayStateTests(@Name("type") DataType type, @Name("valueCount") IntSupplier valueCount, @Name("inOrder") boolean inOrder) {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java:71]: `@Name` makes the logging of failures for parameterized tests much nicer.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java [hunk 5]
```diff
@@ -54,8 +79,9 @@ public ArrayStateTests(DataType type, int valueCount, boolean inOrder) {
             case IP -> ElementType.BYTES_REF;
             default -> throw new IllegalArgumentException();
         };
-        this.valueCount = valueCount;
+        this.valueCount = valueCount.getAsInt();
         this.inOrder = inOrder;
+        logger.info("value count is {}", this.valueCount);
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java:84]: To be removed? It won't flood the logs, but not sure if it's intended

**Review by ivancea — APPROVED** [review]
LGTM! How did you find it? Some other issue?

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/IpArrayState.java
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/IpArrayState.java:1]: I think this is the cause of that one, yeah. I first reproduced it in a serverless test, but I think that's just an artifact of different data layout. Here's the stack trace:
```
[2024-10-25T15:17:51,377][WARN ][r.suppressed             ] [search-DXXSfEi] path: /_query, params: {pretty=true, serverlessRequest=true, operatorRequest=true, error_trace=true}, status: 500 java.lang.ArrayIndexOutOfBoundsException: arraycopy: last source index 64 out of bounds for byte[56]
	at org.elasticsearch.common.util.BigArrays$ByteArrayWrapper.set(BigArrays.java:139)
	at org.elasticsearch.common.util.BytesRefArray.append(BytesRefArray.java:94)
	at org.elasticsearch.compute.data.BytesRefBlockBuilder.appendBytesRef(BytesRefBlockBuilder.java:32)
	at org.elasticsearch.compute.data.BytesRefBlockBuilder.appendBytesRef(BytesRefBlockBuilder.java:20)
	at org.elasticsearch.compute.aggregation.IpArrayState.toIntermediate(IpArrayState.java:130)
	at org.elasticsearch.compute.aggregation.MaxIpAggregator$GroupingState.toIntermediate(MaxIpAggregator.java:91)
	at org.elasticsearch.compute.aggregation.MaxIpGroupingAggregatorFunction.evaluateIntermediate(MaxIpGroupingAggregatorFunction.java:201)
	at org.elasticsearch.compute.aggregation.GroupingAggregator.evaluate(GroupingAggregator.java:75)
	at org.elasticsearch.compute.operator.HashAggregationOperator.finish(HashAggregationOperator.java:201)
	at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:272)
	at org.elasticsearch.compute.operator.Driver.run(Driver.java:189)
```

I tried to write a unit test that was "higher" than this one but gave up after an hour or two.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java (hunk 5 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/ArrayStateTests.java:84]: It's intended though not super important. If the parameter is random it's nice to have the real value written down.

**Pull request merged**